File: snap/snapdir/snapdir.go
// -*- Mode: Go; indent-tabs-mode: t -*-

/*
 * Copyright (C) 2014-2015 Canonical Ltd
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 3 as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

package snapdir

import (
	"fmt"
	"io"
	"os"
	"path/filepath"

	"github.com/ddkwork/golibrary/mylog"
	"github.com/snapcore/snapd/osutil"
	"github.com/snapcore/snapd/snap"
	"github.com/snapcore/snapd/snap/internal"
)

func IsSnapDir(path string) bool {
	if osutil.IsDirectory(path) {
		if osutil.FileExists(filepath.Join(path, "meta", "snap.yaml")) {
			return true
		}
	}
	return false
}

// SnapDir is the snapdir based snap.
type SnapDir struct {
	path string
}

// Path returns the path of the backing container.
func (s *SnapDir) Path() string {
	return s.path
}

// New returns a new snap directory container.
func New(path string) *SnapDir {
	return &SnapDir{path: path}
}

func (s *SnapDir) Size() (size int64, err error) {
	totalSize := int64(0)
	f := func(_ string, info os.FileInfo, err error) error {
		totalSize += info.Size()
		return err
	}
	filepath.Walk(s.path, f)

	return totalSize, nil
}

func (s *SnapDir) Install(targetPath, mountDir string, opts *snap.InstallOptions) (bool, error) {
	// TODO:UC20: support MustNotCrossDevices somehow here

	return false, os.Symlink(s.path, targetPath)
}

func (s *SnapDir) RandomAccessFile(file string) (interface {
	io.ReaderAt
	io.Closer
	Size() int64
}, error,
) {
	f := mylog.Check2(os.Open(filepath.Join(s.path, file)))

	return internal.NewSizedFile(f)
}

func (s *SnapDir) ReadFile(file string) (content []byte, err error) {
	return os.ReadFile(filepath.Join(s.path, file))
}

func (s *SnapDir) ReadLink(file string) (string, error) {
	return os.Readlink(filepath.Join(s.path, file))
}

func (s *SnapDir) Lstat(file string) (os.FileInfo, error) {
	return os.Lstat(filepath.Join(s.path, file))
}

func littleWalk(dirPath string, dirHandle *os.File, dirstack *[]string, walkFn filepath.WalkFunc) error {
	const numSt = 100

	// XXX: check if os.ReadDir is more efficient
	sts := mylog.Check2(dirHandle.Readdir(numSt))

	for _, st := range sts {
		path := filepath.Join(dirPath, st.Name())
		mylog.Check(walkFn(path, st, nil))

		// caller wants to skip this directory

	}

	return nil
}

// Walk (part of snap.Container) is like filepath.Walk, without the ordering guarantee.
func (s *SnapDir) Walk(relative string, walkFn filepath.WalkFunc) error {
	relative = filepath.Clean(relative)
	if relative == "" || relative == "/" {
		relative = "."
	} else if relative[0] == '/' {
		// I said relative, darn it :-)
		relative = relative[1:]
	}
	root := filepath.Join(s.path, relative)
	// we could just filepath.Walk(root, walkFn), but that doesn't scale
	// well to insanely big directories as it reads the whole directory,
	// in order to sort it. This Walk doesn't do that.
	//
	// Also the directory is always relative to the top of the container
	// for us, which would make it a little more messy to get right.
	f := mylog.Check2(os.Open(root))

	defer func() {
		if f != nil {
			f.Close()
		}
	}()

	st := mylog.Check2(f.Stat())
	mylog.Check(walkFn(relative, st, nil))

	if !st.IsDir() {
		return nil
	}

	var dirstack []string
	for {
		if err := littleWalk(relative, f, &dirstack, walkFn); err != nil {
			if err != io.EOF {
				mylog.Check(walkFn(relative, nil, err))
			}
			if len(dirstack) == 0 {
				// finished
				break
			}
			f.Close()
			f = nil
			for f == nil && len(dirstack) > 0 {
				relative = dirstack[0]
				f = mylog.Check2(os.Open(filepath.Join(s.path, relative)))

				dirstack = dirstack[1:]
			}
			if f == nil {
				break
			}
			continue
		}
	}

	return nil
}

func (s *SnapDir) ListDir(path string) ([]string, error) {
	fileInfos := mylog.Check2(os.ReadDir(filepath.Join(s.path, path)))

	var fileNames []string
	for _, fileInfo := range fileInfos {
		fileNames = append(fileNames, fileInfo.Name())
	}

	return fileNames, nil
}

func (s *SnapDir) Unpack(src, dstDir string) error {
	return fmt.Errorf("unpack is not supported with snaps of type snapdir")
}

File: cmd/libsnap-confine-private/snap-dir.c
/*
 * Copyright (C) 2024 Canonical Ltd
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 3 as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

#define _GNU_SOURCE

#include "snap-dir.h"

#include <errno.h>
#include <fcntl.h>
#include <limits.h>
#include <stddef.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <unistd.h>

#include "snap.h"  // For SC_SNAP_MOUNT_DIR_UNSUPPORTED

static const char *_snap_mount_dir = NULL;

// Function is exported only for tests.
void sc_set_snap_mount_dir(const char *dir);
void sc_set_snap_mount_dir(const char *dir) { _snap_mount_dir = dir; }

const char *sc_snap_mount_dir(sc_error **errorp) {
    sc_error *err = NULL;

    if (_snap_mount_dir == NULL) {
        err = sc_error_init_api_misuse("sc_probe_snap_mount_dir_from_pid_1_mount_ns was not called yet");
    }

    sc_error_forward(errorp, err);
    return _snap_mount_dir;
}

void sc_probe_snap_mount_dir_from_pid_1_mount_ns(int root_fd, sc_error **errorp) {
    char target[PATH_MAX] = {0};
    struct stat sb;
    ssize_t n = 0;
    sc_error *err = NULL;

    const char *const probe_dir =
        /* Depending on the value of the root_fd descriptor, the probe path is either absolute or relative.*/
        root_fd == AT_FDCWD
            /* If we are given the special AT_FDCWD descriptor then probe the path "/proc/1/root/snap". */
            ? "/proc/1/root" SC_CANONICAL_SNAP_MOUNT_DIR
            /* If we are given any other descriptor the probe a relative path "proc/1/root/snap",
             * which is relative to root_fd. */
            : "proc/1/root" SC_CANONICAL_SNAP_MOUNT_DIR;

    // If /snap does not exist, then we assume the fallback directory is used.
    if (fstatat(root_fd, probe_dir, &sb, AT_SYMLINK_NOFOLLOW) != 0) {
        if (errno != ENOENT) {
            err = sc_error_init_from_errno(errno, "cannot fstatat canonical snap directory");
            goto out;
        }

        _snap_mount_dir = SC_ALTERNATE_SNAP_MOUNT_DIR;
        return;
    }

    // If /snap exists it must be either a directory or a symbolic link.
    switch (sb.st_mode & S_IFMT) {
        case S_IFDIR:
            _snap_mount_dir = SC_CANONICAL_SNAP_MOUNT_DIR;
            break;

        case S_IFLNK:
            // If /snap exits and is a symbolic link it must point to the fallback directory.
            n = readlinkat(root_fd, probe_dir, target, sizeof target);
            if (n < 0 || n == sizeof(target)) {
                err = sc_error_init_from_errno(errno, "cannot readlinkat canonical snap directory");
                goto out;
            }

            if (strncmp(target, SC_ALTERNATE_SNAP_MOUNT_DIR, n) != 0) {
                err = sc_error_init(SC_SNAP_DOMAIN, SC_SNAP_MOUNT_DIR_UNSUPPORTED,
                                    SC_CANONICAL_SNAP_MOUNT_DIR
                                    " must be a symbolic link to " SC_ALTERNATE_SNAP_MOUNT_DIR);
                goto out;
            }

            _snap_mount_dir = SC_ALTERNATE_SNAP_MOUNT_DIR;
            break;

        default:
            err = sc_error_init(SC_SNAP_DOMAIN, SC_SNAP_MOUNT_DIR_UNSUPPORTED,
                                SC_CANONICAL_SNAP_MOUNT_DIR
                                " must be a directory or a symbolic link to " SC_ALTERNATE_SNAP_MOUNT_DIR);
            goto out;
            break;
    }

out:
    sc_error_forward(errorp, err);
}
